Parse/Parse.m

```diff
@@ -54,17 +54,35 @@ + (void)initialize {
 ///--------------------------------------
 
 + (void)setApplicationId:(NSString *)applicationId clientKey:(NSString *)clientKey {
-    PFConsistencyAssert([applicationId length], @"'applicationId' should not be nil.");
-    PFConsistencyAssert([clientKey length], @"'clientKey' should not be nil.");
-    
-    // Setup new manager first, so it's 100% ready whenever someone sends a request for anything.
-    ParseManager *manager = [[ParseManager alloc] initWithApplicationId:applicationId clientKey:clientKey];
-    [manager configureWithApplicationGroupIdentifier:applicationGroupIdentifier_
-                     containingApplicationIdentifier:containingApplicationBundleIdentifier_
-                               enabledLocalDataStore:shouldEnableLocalDatastore_];
-    currentParseManager_ = manager;
-
+    [self initializeWithConfiguration:[ParseClientConfiguration configurationWithBlock:^(id<ParseMutableClientConfiguration> configuration) {
+        configuration.applicationId = applicationId;
+        configuration.clientKey = clientKey;
+        configuration.localDatastoreEnabled = shouldEnableLocalDatastore_;
+        configuration.applicationGroupIdentifier = applicationGroupIdentifier_;
+        configuration.containingApplicationBundleIdentifier = containingApplicationBundleIdentifier_;
+    }]];
+
+    // TODO (richardross): Investigate why we actually need this here.
     shouldEnableLocalDatastore_ = NO;
+}
+
++ (void)initializeWithConfiguration:(ParseClientConfiguration *)configuration {
+    // Validate configuration
+    PFConsistencyAssert([configuration.applicationId length], @"'applicationId' should not be nil.");
+    PFConsistencyAssert([configuration.clientKey length], @"'clientKey' should not be nil.");
+
+    if (configuration.applicationGroupIdentifier) {
+        if ([PFApplication currentApplication].extensionEnvironment) {
+            PFConsistencyAssert([configuration.containingApplicationBundleIdentifier length], @"'containingApplicationBundleIdentifier' cannot be nil in extension environment");
+        } else {
+            PFConsistencyAssert(![configuration.containingApplicationBundleIdentifier length], @"'containingApplicationBundleIdentifier' must be nil in non-extension environment");
+        }
+
+        PFConsistencyAssert([PFFileManager isApplicationGroupContainerReachableForGroupIdentifier:configuration.applicationGroupIdentifier],
+                            @"ApplicationGroupContainer is unreachable. Please double check your Xcode project settings.");
```

Comment:
IMO most of these checks should go into the builders allocation? Thoughts @nlutsenko @hallucinogen @wangmengyan95?